File: app/schemas/room_type.py
"""
Pydantic schemas for room type management.
"""

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional
from decimal import Decimal


class RoomTypeBase(BaseModel):
    """Base room type schema with common fields."""
    name: str = Field(..., min_length=1, max_length=100, description="Room type name")
    total_rooms: int = Field(..., gt=0, description="Total number of rooms of this type")
    base_price: Decimal = Field(..., gt=0, description="Base price per night")
    description: Optional[str] = Field(None, max_length=500, description="Room type description")


class RoomTypeCreate(RoomTypeBase):
    """Schema for creating a new room type."""
    pass


class RoomTypeUpdate(BaseModel):
    """Schema for updating a room type."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    total_rooms: Optional[int] = Field(None, gt=0)
    base_price: Optional[Decimal] = Field(None, gt=0)
    description: Optional[str] = Field(None, max_length=500)


class RoomTypeRead(RoomTypeBase):
    """Schema for reading room type data."""
    id: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

File: frontend/src/services/roomType.ts
import { api } from '../lib/axios';

export interface RoomType {
    id: number;
    name: string;
    description?: string;
    total_rooms: number;
    base_price: number;
    created_at: string;
    updated_at: string;
}

export interface RoomTypeCreate {
    name: string;
    description?: string;
    total_rooms: number;
    base_price: number;
}

export const roomTypeService = {
    getAll: async () => {
        const response = await api.get<RoomType[]>('/room-types');
        return response.data;
    },

    getById: async (id: number) => {
        const response = await api.get<RoomType>(`/room-types/${id}`);
        return response.data;
    },

    create: async (data: RoomTypeCreate) => {
        const response = await api.post<RoomType>('/room-types', data);
        return response.data;
    },

    update: async (id: number, data: Partial<RoomTypeCreate>) => {
        const response = await api.put<RoomType>(`/room-types/${id}`, data);
        return response.data;
    },

    delete: async (id: number) => {
        await api.delete(`/room-types/${id}`);
    },
};
